File: solutions/0003_longest-substring-without-repeating-characters.java
// Problem: https://leetcode.com/problems/longest-substring-without-repeating-characters

class Solution {
    HashMap<Character, Integer> lookUpTable = new HashMap<>(); 

    public int lengthOfLongestSubstring(String s) {
        int length = 0;
        int longest = 0;
        for (int i = 0; i < s.length(); i++) {
            int lastOcc = lookUpTable.getOrDefault(s.charAt(i), -1);
            lookUpTable.put(s.charAt(i), i);
            if (i - lastOcc > length) {
                // this char won't be duplicate
                length += 1;
                longest = length > longest ? length : longest; // max
            } else {
                length = i - lastOcc;
            }
        }

        return longest;
    }
}

File: solutions/0003_longest-substring-without-repeating-characters.py
# Problem: https://leetcode.com/problems/longest-substring-without-repeating-characters
# Runtime: 34 ms

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        lastSeenTable = {}
        length = 0
        longest = 0
        for i, char in enumerate(s):
            lastSeen = lastSeenTable.get(char, -1)
            lastSeenTable[char] = i
            if (i - lastSeen > length):
                length += 1
                longest = max(length, longest)
            else:
                length = i - lastSeen
        return longest